forward type alignments to allocators

krux02 opened the pull request:
* Introduce ``NIM_ALIGNOF`` in `nimbase.h` in the style of `NIM_ALIGN`.
 * Add alignment to RTTI.
 * Forward all alignment constraints of allocations to the allocator (allocator does not support custom alignment yet).
 * Fixes a problem with SIMD types in lambda lifting.
 * This is some foundation work to solve #7865.
 * Introduce define ``nimAlignPragma`` for bootstrapping (original alignas PR  didn't need it).

This PR was previeously knows as ``Fix for newSeq on Aligned Type``, since that part isn't solved, I renamed it.

When the tests are passing, this PR should be ready for review (and possibly merge).

cooldome commented:
I have tested thing one and I find it very use. It makes alignment information to flow though all the levels way down to allocator. Allocator is not yet alignment aware but it can be fixed in next PR.

One more issue needs to be addressed. Closure environments needs to aligned too when they capture a variable of aligned type

cooldome commented:
Test case for closures. Compile in release mode.
```nim
type
  m128d {.importc: "__m128d", header: "immintrin.h".} = object

proc add(a: m128d; b: m128d): m128d {.importc: "_mm_add_pd", header: "immintrin.h".}
proc set1*(a: float): m128d {.importc: "_mm_set1_pd", header: "immintrin.h".}
func `+`(a,b: m128d): m128d = add(a, b)


proc lambdaGen(a, b: float) : auto =
  let x1 = set1(2.0 + a)
  let x2 = set1(-23.0 - b)
  let capturingLambda = proc(x: m128d): m128d =
    let cc = x1 + x1
    let bb = x2 + set1(12.5)
    result = cc + bb + x
  return capturingLambda

let f1 = lambdaGen(2.0 , 2.221)
let f2 = lambdaGen(-1.226 , 3.5)

echo f1(set1(2.0))
echo f2(set1(-23.0
```

Araq commented on lib/system.nim:
```diff
@@ -674,6 +674,7 @@ when not defined(JS) and not defined(nimSeqsV2):
       len, reserved: int
       when defined(gogc):
         elemSize: int
+        elemAlign: int
```
I doubt Go's layout rules for seqs care about ``elemAlign``.

krux02 replied:
It is used here: https://github.com/krux02/Nim/blob/fc6965e0487aa328efaaef6a3830b3caec7a7e62/lib/system/mmdisp.nim#L331

Araq commented on lib/system/alloc.nim:
```diff
@@ -65,16 +64,20 @@ type
+    when defined(nimAlignPragma):
+      data {.align: MemAlign.}: UncheckedArray[byte]      # start of usable memory
+    else:
+      data: UncheckedArray[byte]
```
Dangerous for bootstrapping purposes. Why not keep the old, more correct variant for it?

krux02 replied:
Yea, I thought that you might comment on this part. Well, I changed the value for `MemAlign` from 8 to 16, this is also alignment guarantee for malloc in C. This will also work for many SIMD data types if native optimizations are not turned on.

Why not keep the old, more correct variant? Well the old variant is with the new value of `MemAlign` not correct anymore. It is true that alignment constraints aren't met during bootstrapping. But the question is: "Are the alignment constraints necessary during compilation?", and I would say: "No". There are no SIMD types used in the compiler yet. And if there are ever SIMD types used in the compiler, the branch without the align pragme should be dead by then.
I also wouldn't call it "dangerous", everything that might happen is, during Bootstrapping on 32 bit systems, float64 operations might be not as fast as they could be.

Araq commented on lib/system/mmdisp.nim:
```diff
@@ -46,7 +46,7 @@ const
   PageMask = PageSize-1
 
-  MemAlign = 8 # also minimal allocatable memory block
+  MemAlign = 16 # also minimal allocatable memory block
```
This adds quite some overhead to the Nim compiler itself. The compiler doesn't care about SSE data anyway. Not every piece of software is a game.

krux02 replied:
I explained earlier why I changed the value. If you really really don't want it, I can revert it back. But it basically means that SSE can't be used at all. Not every software that needs SSE is a game.

cooldome replied:
I suggest the following approach. 
This PR gives us automatic flow of alignment type info. Now we can do dynamic alignment in allocator instead of static 8 or 16 bytes. This is much better than C/C++ approach that always overallocates to 16 bytes in 64bit mode because malloc has no access to type info.

For example if struct consists only of chars then it will get alignment of 1 and allocator can take advantage of this info and provide most efficient in terms memory usage allocation. 
If object contains SSE it will get 16 byte alignment, if AVX it will get 32 byte alignment. 
`MemAlign` no longer needs to be a constant.

I suggest passing the alignment as argument to allocator. In first implementation it can still overallocate upwards. I suggest allocator to have API similar to this one https://github.com/microsoft/mimalloc/blob/acb03c54971c4b0a43a6d17ea55a9d5feb88972f/src/alloc-aligned.c#L17

```nim
proc rawAllocAlignedAt(a: var MemRegion, size: int, offset = 0, align = MemAlign): pointer =
```
In this way allocator can allocate aligned, offset aligned (required because of FreeCell header).

Araq commented on tests/seq/thardalignmentconstraint.nim:
```diff
@@ -0,0 +1,17 @@
+disabled: true
+"""
+
+# does not yet work
```
Well the tests are green now, at least.

krux02 replied:
It is disabled. Maybe I should remove it since you don't like disabled tests.

krux02 commented:
@cooldome you example is incomplete, but it got the point. I could verify it on my end that the example worked with this PR applied. But I can't explain why, I didn't do anything in lambda lifting. All I do is forwarding the alignment information to the allocators. When I added your example to the tests, it failed on 32 bit systems. So I wouldn't say I fixed anything in that direction. But I think this PR will help to make it work in the future.

cooldome commented:
Somehow I missed that a big progress in this PR. Starting active testing.

@krux02: in 32bit systems the problem that stack it self is not 16 byte aligned hence fixing allocator doesn't safe a day. I gave up SSE vectorization in 32bit mode even in C++. I do the following builds these days:  32 bit unvectorised, 64bit SSE, 64bit AVX, 64bit AVX2. Fixing 32bit would require your own calling convention which out of scope of this PR. This calling convention should be used for proc and closures.

I don't think you need to do anything special in lambda lifting. LL defines the environment object and as long as `new(MyEnvObject)` is aligned then everything should work

cooldome commented:
I have tried it, looks like SSE with 16 bytes alignment now works, but not the AVX that needs 32 byte alignment. I guess it boils down to allocator that still have one fixed alignment,

krux02 commented:
This was very painful to merge with the development branch. I won't do it again. Either merge it or reject it.

Araq commented:
Irrelevant CI failure, merging.
